validator/keymanager/remote-web3signer/keymanager.go

```diff
@@ -38,54 +53,323 @@ type SetupConfig struct {
-func NewKeymanager(_ context.Context, cfg *SetupConfig) (*Keymanager, error) {
+func NewKeymanager(ctx context.Context, cfg *SetupConfig) (*Keymanager, error) {
+	ctx, span := trace.StartSpan(ctx, "remoteKeymanager")
+	defer span.End()
 	if cfg.BaseEndpoint == "" || !bytesutil.IsValidRoot(cfg.GenesisValidatorsRoot) {
 		return nil, fmt.Errorf("invalid setup config, one or more configs are empty: BaseEndpoint: %v, GenesisValidatorsRoot: %#x", cfg.BaseEndpoint, cfg.GenesisValidatorsRoot)
 	}
 	client, err := internal.NewApiClient(cfg.BaseEndpoint)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not create apiClient")
 	}
-	return &Keymanager{
+
+	km := &Keymanager{
 		client:                internal.HttpSignerClient(client),
 		genesisValidatorsRoot: cfg.GenesisValidatorsRoot,
 		accountsChangedFeed:   new(event.Feed),
-		publicKeysURL:         cfg.PublicKeysURL,
-		providedPublicKeys:    cfg.ProvidedPublicKeys,
 		validator:             validator.New(),
-		publicKeysUrlCalled:   false,
-	}, nil
-}
+		retriesRemaining:      maxRetries,
+		keyFilePath:           cfg.KeyFilePath,
+	}
 
-// FetchValidatingPublicKeys fetches the validating public keys
-// from the remote server or from the provided keys if there are no existing public keys set
-// or provides the existing keys in the keymanager.
-func (km *Keymanager) FetchValidatingPublicKeys(ctx context.Context) ([][fieldparams.BLSPubkeyLength]byte, error) {
-	if km.publicKeysURL != "" && !km.publicKeysUrlCalled {
-		providedPublicKeys, err := km.client.GetPublicKeys(ctx, km.publicKeysURL)
+	keyFileExists := false
+	if km.keyFilePath != "" {
+		keyFileExists, err = file.Exists(km.keyFilePath, file.Regular)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not check if remote signer persistent keys exists in %s", km.keyFilePath)
+		}
+		if !keyFileExists {
+			return nil, fmt.Errorf("no file exists in remote signer key file path %s", km.keyFilePath)
+		}
+	}
+
+	var ppk []string
+	// load key values
+	if cfg.PublicKeysURL != "" {
+		providedPublicKeys, err := km.client.GetPublicKeys(ctx, cfg.PublicKeysURL)
 		if err != nil {
 			erroredResponsesTotal.Inc()
-			return nil, errors.Wrap(err, fmt.Sprintf("could not get public keys from remote server url: %v", km.publicKeysURL))
+			return nil, errors.Wrapf(err, "could not get public keys from remote server url: %v", cfg.PublicKeysURL)
+		}
+		ppk = providedPublicKeys
+	} else if len(cfg.ProvidedPublicKeys) != 0 {
+		ppk = cfg.ProvidedPublicKeys
+	}
+
+	// use a map to remove duplicates
+	flagLoadedKeys := make(map[string][48]byte)
+
+	// Populate the map with existing keys
+	for _, key := range ppk {
+		decodedKey, err := hexutil.Decode(key)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not decode public key for remote signer")
+		}
+		if len(decodedKey) != fieldparams.BLSPubkeyLength {
+			return nil, fmt.Errorf("invalid public key length: expected %d, got %d", fieldparams.BLSPubkeyLength, len(decodedKey))
```

Comment:
```suggestion
			return nil, fmt.Errorf("public key %s has invalid length (expected %d, got %d)", decodedKey, fieldparams.BLSPubkeyLength, len(decodedKey))
```
Error messages are separated by colons so having a colon as part of an error is confusing.